Add Confluent provisioner support

dhilst opened the pull request.

dhilst committed:
Update .gitignore

dhilst committed:
HOTFIX some bugs found during the installation

- Fix exit code unhandling bug
- Get kernel version from answerfile
- Fix OFED version in repos.conf
- Fix image generation with custom kernel
- Fix spack idempotency
- Fix diskImage use before set

dhilst committed:
Run clang-format

dhilst committed:
HOTFIX Enable keepcache in dnf

dhilst committed:
HOTFIX: Fix from where the kernel version is fetched

viniciusferrao committed:
Fix unreachable code in fetchSubnetMask

dhilst committed:
Use running kernel when kernel option is ommited in the answerfile

- In that case run genimage without customizing the kernel image

dhilst committed:
Fix slurld error in the computing node

dhilst committed:
Comment kernel in rocky9-base.ini

dhilst committed:
Fix makedns command & restart chronyd

dhilst committed:
Generalize ScriptBuilder tests & fix warnings

dhilst committed:
Added TODOs for confluent integration

dhilst commented on src/services/ansible/roles/base.cpp:
```diff
@@ -60,6 +60,7 @@ ScriptBuilder installScript(
 
     // "python3-dnf-plugin-versionlock" is conflicting with dnf-plugins-core
     // during the first install
+    // TODO: CFL initscripts is only required by xCAT
```
`initscripts` should be moved to `xCAT.cpp` again

viniciusferrao commented on src/services/scriptbuilder.cpp:
```diff
@@ -96,15 +96,10 @@ TEST_CASE("Basic")
-  echo "123.123.123.123 example.com" >> "/etc/hosts"
-systemctl enable --now foo-service)del");
+    const auto script = builder.toString();
+    CHECK(script.contains(R"_(echo "123.123.123.123 example.com" >> "/etc/hosts")_"));
```
Use a reserved address for documentation always. You can use any of the three TEST-NETS available on [RFC5735](https://www.rfc-editor.org/rfc/rfc5735#page-6).

`192.0.2.0/24` it's a good example.

viniciusferrao commented on test/network.cpp:
```diff
@@ -7,8 +7,13 @@
-
+    TEST_CASE("fetchSubnetMask throws for unknown interface")
+    {
+        CHECK_THROWS_AS(static_cast<void>(Network::fetchSubnetMask("nonexistent0")),
```
Do we need this `cast`?

viniciusferrao commented on src/main.cpp:
```diff
@@ -60,6 +61,9 @@ int runTestCommand(const std::string& testCommand,
         LOG_INFO("Groups: {}", fmt::join(file.getGroups(), ","));
         LOG_INFO("Contents: {}", file.toData());
+    } else if (testCommand == "confluent-install") {
+        Confluent cfl;
```
Should we use the `Provisioner` class instead?

viniciusferrao commented on src/NFS.cpp:
```diff
@@ -169,46 +143,20 @@ TEST_CASE("installImageScript")
+    CHECK(script.contains("systemctl enable autofs"));
+    CHECK(script.contains(R"(echo "nfs-utils" >> "${PKGLIST}")"));
+    CHECK(script.contains(R"(echo "autofs" >> "${PKGLIST}")"));
+    CHECK(script.contains(R"(echo "/home   /etc/auto.home" >> "${ROOTFS}/etc/auto.master")"));
+    CHECK(script.contains(R"(echo "/opt/ohpc/pub   /etc/auto.ohpc" >> "${ROOTFS}/etc/auto.master")"));
```
Use `\t`.

viniciusferrao commented on include/cloysterhpc/services/provisioner.h:
```diff
@@ -8,6 +8,8 @@
 
 #include <string>
 
+// TODO: CFL Make this a generic interface between XCAT and Confluent
```
CFL?

viniciusferrao commented on src/services/shell.cpp:
```diff
@@ -445,6 +446,9 @@ void Shell::install()
     // will output a lot of helpful information in the logs
     dumpPreInstallState();
 
+    // Initialize repositories. This will setup Rocky Vault and assemble
```
We should be distribution agnostic. Vault repositories should be an specialisation of the OS.

dhilst committed:
General code improvements and fixes

* Make kernel optional from the answerfile and do not
  use the kernel version if it is ommited, this includes
  xCAT and OFED modules. When the version is required
  the output of uname -r is being used
* Add some file utilities in cloyster::services::files namespace
* Add a cache module in cloyster::services::cache::fs that run
  a general function and caches the result in the disk, or returns
  the cached value
* Add cloyster::services::cache::fs::checksum to cache the results
  of checksums in the disk
* Use cloyster::services::cache::fs::checksum in diskImage to
  cache the results of the ISO checksum
* Fix disk image checksum fault
* Remove dnf command from answerfile

dhilst committed:
WIP Restructuring the code to enable modularity

dhilst committed:
Wrap ifaddrs pointers

dhilst committed:
Make Answerfile a singleton

dhilst committed:
Add ansible roles boilerplate

dhilst committed:
Migrating the logic to ansible roles

dhilst committed:
Add ansible Execution class

dhilst committed:
Implement provisioner role and finish Shell -> roles migration

dhilst committed:
Add Confluent install logic sketch

dhilst committed:
Make roles interface type safer

dhilst committed:
Make provisioner polymorphic at compile time

dhilst committed:
Enable setup network code to continue to run in debug buids

dhilst committed:
Use singleton getters and enforce const correcteness of some singletons

dhilst committed:
Log the intialization and split network and sshd roles

viniciusferrao commented on src/services/cache.cpp:
```diff
@@ -0,0 +1,12 @@
+
+std::string checksum(std::string_view name, const std::filesystem::path& path)
+{
+    auto wrapper = [&]() -> std::string {
+        return files::checksum(path);
+    };
```
There's no reason for this lambda. Just return.

dhilst committed:
Test repos role

dhilst committed:
Small fixes in the ansible roles

dhilst committed:
Run NFS and Confluent roles

dhilst committed:
Define nodes in confluent

- Add container utility functions

dhilst committed:
Confluent deployment fixes

dhilst committed:
Fix Confluent deployment
